### PR TITLE
chore: Performance fix generateDataTreeWidgetWithoutMeta

### LINKED ISSUE #31267: getAllPathsFromPropertyConfigWithoutMemo performance fix
We have observed this function getAllPathsFromPropertyConfigWithoutMemo is very slow, we have made some optimisations in the merge and the hideColumnType.

### PR DESCRIPTION
**Pull request by vsvamsi1**
## Description
Made code optimisations around spread and the hideColumnType computation logic, since the combined latency of these operations were showing up significantly in the performance profile.
#### PR fixes following issue(s)
Fixes #31267

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)
- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
	- Introduced a new configuration option in the Table Widget to hide properties based on column types.
	- Updated state management in widgets using `immer` for better performance and reliability.

### CODE REVIEW

File: app/client/src/entities/Widget/utils.ts [hunk 1]
```diff
@@ -125,22 +130,23 @@ const childHasPanelConfig = (
...
+                        );
+                        mergeObjectThroughMutation(
+                          triggerPaths,
+                          configTriggerPaths,
+                        );
+                        mergeObjectThroughMutation(
+                          validationPaths,
+                          configValidationPaths,
+                        );
+
```
Comment (vsvamsi1) [inline app/client/src/entities/Widget/utils.ts:149]: @riodeuno @ashit-rath The merge orders within `childHasPanelConfig` function is functionally different from `getAllPathsFromPropertyConfigWithoutMemo`, the target object has more precedence over here which seems strange.

There are no issues when the keys are unique during these merges but if they are the same then the target object has more precedence.

 Is this merge behaviour intentional in the `childHasPanelConfig`? if it is then i have to change the way i merge from this to 
`
const mergeObjectThroughMutation = (draft: any, obj: object) =>
  Object.entries(obj).forEach(([key, value]) => {
    draft[key] = value;
  });
`
to a logic like this 
`
const mergeObjectThroughMutationFlip = (draft: any, obj: object) =>
  Object.entries(obj).forEach(([key, value]) => {
    if (!(key in draft)) {
      draft[key] = value;
    }
  });
`
to maintain functional parity over here. Is this required?

cc @rajatagrawal

**Comment by rajatagrawal** [comment]
@vsvamsi1 

The changes in this [PR](https://github.com/appsmithorg/appsmith/pull/31268) seem to further improve the performance of the generateDataTreeWidget function.

1. The recursive approach seems to reduce ~200ms per execution of `childHasPanelConfig` function.
2. Memoization of `childHasPanelConfig` function reduces the overall time taken by `DataTreeFactory.create` function.

Would you like to consider/review or incorporate any part of the changes in your PR ?

**Review by rajatagrawal — APPROVED** [review]

**Pull request merged**